python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -1430,10 +1432,57 @@ def delete(self, table: sqlalchemy.schema.Table, columns: Iterable[str], *rows:
             # while reporting that no rows were affected.
             return 0
         sql = table.delete()
-        whereTerms = [table.columns[name] == sqlalchemy.sql.bindparam(name) for name in columns]
-        if whereTerms:
-            sql = sql.where(sqlalchemy.sql.and_(*whereTerms))
-        return self._connection.execute(sql, *rows).rowcount
+        columns = list(columns)  # Force iterators to list
+
+        # More efficient to use IN operator if there is only one
+        # variable changing across all rows.
+        content: Dict[str, Set] = defaultdict(set)
+        if len(columns) == 1:
+            # Nothing to calculate since we can always use IN
+            column = columns[0]
+            changing_columns = [column]
+            content[column] = set([row[column] for row in rows])
+        else:
+            for row in rows:
+                for k, v in row.items():
+                    content[k].add(v)
+            changing_columns = [col for col in content if len(content[col]) > 1]
+
+        if len(changing_columns) == 1:
+            name = changing_columns.pop()
+
+            # Simple where clause for the unchanging columns
+            clauses = []
+            for k, v in content.items():
+                if k == name:
+                    continue
+                column = table.columns[k]
+                # The set only has one element
+                clauses.append(column == v.pop())
+
+            # The IN operator will not work for "infinite" numbers of
+            # rows so must batch it up into distinct calls.
+            in_content = list(content[name])
+            n_elements = len(in_content)
+            MAX_ELEMENTS_PER_IN = 1500
+            n_loops = math.ceil(n_elements / MAX_ELEMENTS_PER_IN)
+            n_per_loop = math.ceil(n_elements / n_loops)
```

Comment:
This seems kind of complicated; why not just take MAX_ELEMENTS_PER_IN at a time until you run out?  Then (because you would explicitly check) you also don't have the possibility that endpos is off the end (which I think you have now).

Reply:
I was trying not to have a situation where there are 5001 rows and 1000 as MAX_ELEMENTS_PER_IN to make it run 6 times when I can make it run 5 times if I use 1001.

Reply:
I do realize the logic is not quite right regardless but the loop still covers everything.

Reply:
Except I think you still get 6 loops of 834 in the case you gave, not 5 of 1001.  What's the use of having a MAX if it's not the maximum?  And is an extra loop/query really that much time (compared with the thousands that you are avoiding)?

Reply:
Just saw you have a new calculation.  It is better, but still a lot of comments and code for not a lot of gain.

Reply:
Yes. Whilst doing the school run I realized which way I had messed it up. I can remove the complication -- I just felt like I preferred a more even spread...